### PR TITLE
Add Kyber Governance interfaces and logics, support binary and generic proposals

### PR DESCRIPTION
**Pull request by manhlx3006**

### CODE REVIEW

File: contracts/governance/KyberGovernance.sol [hunk 1]
```diff
@@ -234,64 +312,84 @@ contract KyberGovernance is Ownable, IKyberGovernance {
...
+  ) external override {
+    for(uint256 i = 0; i < proposalIds.length; i++ ) {
+      if (getProposalState(proposalIds[i]) == ProposalState.Active) {
+        ProposalWithoutVote storage proposal = _proposals[proposalIds[i]];
+        require(address(proposal.strategy) == msg.sender, "ONLY_VOTING_POWER_STRATEGY");
+        Vote memory vote = proposal.votes[staker];
+        if (vote.choice == 0) continue; // not voted yet
+        uint256 oldVotingPower = uint256(vote.votingPower);
+        proposal.totalVotes = proposal.totalVotes + newVotingPower - oldVotingPower;
+        // update vote count for each voted option
```
Comment (ducquangkstn) [inline contracts/governance/KyberGovernance.sol:372]: add a comment that proposal is binary then `proposal.options.length` is 2, and choice is 1 or 2

File: contracts/governance/KyberGovernance.sol [hunk 2]
```diff
@@ -234,64 +312,84 @@ contract KyberGovernance is Ownable, IKyberGovernance {
...
+        for(uint256 j = 0; j < proposal.options.length; j++) {
+          if (vote.choice & (2 ** j) == (2 ** j)) {
+            // expect proposal.voteCounts[i] >= oldVotingPower
+            proposal.voteCounts[j] = proposal.voteCounts[j] + newVotingPower - oldVotingPower;
+          }
+        }
+        // update voting power of the staker
+        _proposals[proposalIds[i]].votes[staker] = Vote({
```
Comment (ducquangkstn) [inline contracts/governance/KyberGovernance.sol:380]:
```solidity
_proposals[proposalIds[i]].votes[staker].votingPower == uint224(newVotingPower)
```

Comment (ducquangkstn) [reply contracts/governance/KyberGovernance.sol:380]: Pls also add safeUint224 function

**Review by ducquangkstn — COMMENTED** [review]
Do we use SafeMath for KyberGovernance

File: contracts/governance/KyberGovernance.sol (hunk 2 above)
Comment (manhlx3006) [reply contracts/governance/KyberGovernance.sol]: Added safe uint 224 and safe uint 32

File: contracts/governance/KyberGovernance.sol [hunk 3]
```diff
@@ -234,64 +312,84 @@ contract KyberGovernance is Ownable, IKyberGovernance {
...
    **/
-  function setVotingDelay(uint256 votingDelay) external override onlyOwner {
-    _setVotingDelay(votingDelay);
+  function handleVotingPowerChanged(
+    address staker,
+    uint256 newVotingPower,
+    uint256[] calldata proposalIds
+  ) external override {
+    for(uint256 i = 0; i < proposalIds.length; i++ ) {
+      if (getProposalState(proposalIds[i]) == ProposalState.Active) {
```
[the remark this answers is not in the document]
Comment (manhlx3006) [reply contracts/governance/KyberGovernance.sol]: alr changed in the latest version

File: contracts/governance/KyberGovernance.sol (hunk 1 above)
Comment (manhlx3006) [reply contracts/governance/KyberGovernance.sol]: it is added in the submitVote, here is handleVotingPowerChanges, so it shouldn't care about that

**Review by Anyhowclick — CHANGES_REQUESTED** [review]
Require comment standardization:
- Single quote `'` or double quotes `"`?
- `FULL_CAPS` or `low caps`

File: contracts/governance/KyberGovernance.sol [hunk 4]
```diff
@@ -2,197 +2,288 @@
...
+    ProposalWithoutVote storage proposal = _proposals[proposalId].proposalData;
     require(
-      msg.sender == _guardian ||
-        IProposalValidator(address(proposal.executor)).validateProposalCancellation(
-          this,
-          proposal.creator,
-          block.number - 1
-        ),
-      'PROPOSITION_CANCELLATION_INVALID'
+      msg.sender == _daoOperator ||
```
[the remark this answers is not in the document]
Comment (manhlx3006) [reply contracts/governance/KyberGovernance.sol:246]: should be here, to make sure we can always cancel a proposal by daoOperator

**Review by Anyhowclick — APPROVED** [review]

**Review by ducquangkstn — APPROVED** [review]

**Pull request merged**